uber/models.py

```diff
@@ -1625,6 +1817,464 @@ def must_contact(self):
+        """
+        Groups a list of words by their part_of_speech.
+
+        Arguments:
+            words (list): List of `PromoCodeWord`.
+
+        Returns:
+            OrderedDict: A dictionary of words mapped to their part of speech,
+                like this::
+
+                    OrderedDict([
+                        (0, ['adjective1', 'adjective2']),
+                        (1, ['noun1', 'noun2']),
+                        (2, ['verb1', 'verb2']),
+                        (3, ['adverb1', 'adverb2'])
+                    ])
+        """
+        parts_of_speech = OrderedDict(
+            [(i, []) for (i, _) in PromoCodeWord.PART_OF_SPEECH_OPTS])
+        for word in words:
+            parts_of_speech[word.part_of_speech].append(word.word)
+        return parts_of_speech
+
+    @classmethod
+    def normalize_word(cls, word):
+        """
+        Normalizes a word.
+
+        Arguments:
+            word (str): A word as typed by an admin.
+
+        Returns:
+            str: A copy of `word` converted to all lowercase, and multiple
+                whitespace characters replaced by a single space.
+        """
+        return re.sub(r'\s+', ' ', word.strip().lower())
+
+
+class PromoCode(MagModel):
+    """
+    Promo codes used by attendees to purchase badges at discounted prices.
+
+    Attributes:
+        code (str): The actual textual representation of the promo code. This
+            is what the attendee would have to type in during registration to
+            receive a discount. `code` may not be an empty string or a string
+            consisting entirely of whitespace.
+        discount (int): The discount amount that should be applied to the
+            purchase price of a badge. The interpretation of this value
+            depends on the value of `discount_type`. In any case, a value of
+            0 equates to a full discount, i.e. a free badge.
+        discount_display (str): A human readable description of the discount.
+        discount_type (int): The type of discount this promo code will apply.
+            Valid values are:
+
+            * 0 `FIXED_DISCOUNT`: `discount` is interpreted as a fixed
+                dollar amount by which the badge price should be reduced. If
+                `discount` is 49 and the badge price is normally $100, then
+                the discounted badge price would be $51.
+
+            * 1 `FIXED_PRICE`: `discount` is interpreted as the actual badge
+                price. If `discount` is 49, then the discounted badge price
+                would be $49.
+
+            * 2 `PERCENT_DISCOUNT`: `discount` is interpreted as a percentage
+                by which the badge price should be reduced. If `discount` is
+                20 and the badge price is normally $50, then the discounted
+                badge price would $40 ($50 reduced by 20%). If `discount` is
+                100, then the price would be 100% off, i.e. a free badge.
```

Comment:
That is true, for percent discounts both `0` and `100` are a free badge. I believe this is acceptable because we've already decided that we're not allowing promo codes that do not change the price of a badge. So a %0 discount is an invalid value.

Additionally, a `0` dollar fixed discount is also considered a free badge, for the same reasons as above.